sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/CSVInferSchema.scala

```diff
@@ -104,6 +108,7 @@ class CSVInferSchema(val options: CSVOptions) extends Serializable {
           compatibleType(typeSoFar, tryParseDecimal(field)).getOrElse(StringType)
         case DoubleType => tryParseDouble(field)
         case TimestampType => tryParseTimestamp(field)
+        case DateType => tryParseDate(field)
```

Comment:
Another problem here is the order here matters when type is being merged. For instance, date type is inferred first, and then if timestamp is found, it won't detect timestamp type anymore.

Reply:
IIRC we decided to follow the order in partition inference and infer timestamp first?

Reply:
Here's what I mean:

```scala
Seq("2010|10|10", "2010_10_10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: string (nullable = true)
```

```scala
Seq("2010_10_10", "2010|10|10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: date (nullable = true)
```

```scala
Seq("2010_10_10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("timestampFormat", "yyyy_MM_dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: timestamp (nullable = true)
```

```scala
Seq("2010|10|10")
  .toDF.repartition(1).write.mode("overwrite").text("/tmp/foo")
spark.read
  .option("inferSchema", "true")
  .option("header", "false")
  .option("dateFormat", "yyyy|MM|dd").csv("/tmp/foo").printSchema()
```

```
root
 |-- _c0: date (nullable = true)

```

Reply:
It's okay .. sorry for rushing comments. I realised my comments are hard to read now.